## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [3]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [137]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'subtotal', 'shipping_price', 'total_price', 'cost',
       'cost_of_each_line', 'profit_of_each_line'],
      dtype='object')

In [5]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [6]:
# Use this space to do any additional research
# and familiarize yourself with the data.

#Product Analysis
average_price = df['unit_price'].mean()
average_cost = df['unit_cost'].mean()
average_weight = df['unit_weight'].mean()
total_units = df['qty'].sum()
cost_per_unit = average_cost / total_units
average_weight_per_unit = total_units / average_weight
print(f"Product Analysis (Average):\n"
      f"Price of product: ${round(average_price, 2)}\n"
      f"Cost of product: ${round(average_cost, 2)}\n"
      f"Weight of product: {round(average_weight, 2)} lbs\n"
      f"Cost per Pound: ${round(average_cost / average_weight, 2)} per pound")


Product Analysis (Average):
Price of product: $136.27
Cost of product: $99.45
Weight of product: 5.0 lbs
Cost per Pound: $19.87 per pound


In [7]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [24]:
# For the category with the most entries, which subcategory had the most entries?
most_entries = 'consumables'
filtered_df = df[df['category'] == most_entries]
filtered_df['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [43]:
# Which five clients had the most entries in the data?
client_most_entries = df[['first', 'last']].value_counts().head(5)
client_most_entries

first      last   
Jessica    Reyes      220
Angela     Everett    211
Bryan      Myers      209
Alexandra  Young      207
Kendra     Garrett    207
Name: count, dtype: int64

In [40]:
# Store the client ids of those top 5 clients in a list.
top_5_list = df['client_id'].value_counts().head(5).index.tolist()
top_5_list

[33615, 66037, 46820, 38378, 24741]

In [47]:
# How many total units (the qty column) did the client with the most entries order order?
qty_most_ordered = df['qty'].value_counts().head(1)
qty_most_ordered

qty
24    607
Name: count, dtype: int64

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [48]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [138]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df['unit_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
df[['first', 'last', 'unit_weight', 'shipping_price']]

first       last  unit_weight  shipping_price
0         Donald    Harding         7.50            75.0
1        Tiffany      Myers         1.49            14.9
2        Shannon     Watson         1.68            16.8
3         Nathan      Baker         1.23            12.3
4      Christina   Schwartz        46.43           464.3
...          ...        ...          ...             ...
54634     Lauren      Reese         2.25            22.5
54635    Derrick      Moore        11.70           117.0
54636     Monica  Gutierrez         4.16            41.6
54637      Wanda    Solomon        18.04           180.4
54638     Dwayne       Bell         2.07            20.7

[54639 rows x 4 columns]

In [162]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax = (df['shipping_price'] + df['subtotal']) * .0925
df['total_price'] = round(df['shipping_price'] + df['subtotal'] + sales_tax, 2)




In [163]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['cost_of_each_line'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df[['first', 'last', 'unit_cost', 'qty', 'shipping_price', 'cost_of_each_line']]

first       last  unit_cost   qty  shipping_price  \
0         Donald    Harding     762.71   105            75.0   
1        Tiffany      Myers      15.09    21            14.9   
2        Shannon     Watson       7.86    39            16.8   
3         Nathan      Baker      24.85    29            12.3   
4      Christina   Schwartz     108.17    20           464.3   
...          ...        ...        ...   ...             ...   
54634     Lauren      Reese      51.60    33            22.5   
54635    Derrick      Moore     175.46    47           117.0   
54636     Monica  Gutierrez      57.31   475            41.6   
54637      Wanda    Solomon       1.22   112           180.4   
54638     Dwayne       Bell       1.61  1031            20.7   

       cost_of_each_line  
0               80159.55  
1                 331.79  
2                 323.34  
3                 732.95  
4                2627.70  
...                  ...  
54634            1725.30  
54635            8363.62  
54636           27263.85  
54637             317.04  
54638            1680.61  

[54639 rows x 6 columns]

In [156]:
# Create a column for the profit of each line using line cost and line price
df['profit_of_each_line'] = (df['unit_price'] * df['qty']) - df['cost_of_each_line']
df[['first', 'last', 'unit_price', 'cost_of_each_line', 'profit_of_each_line']]


first       last  unit_price  cost_of_each_line  \
0         Donald    Harding     1096.80           80159.55   
1        Tiffany      Myers       24.95             331.79   
2        Shannon     Watson       13.52             323.34   
3         Nathan      Baker       36.42             732.95   
4      Christina   Schwartz      195.10            2627.70   
...          ...        ...         ...                ...   
54634     Lauren      Reese       83.13            1725.30   
54635    Derrick      Moore      206.59            8363.62   
54636     Monica  Gutierrez       65.66           27263.85   
54637      Wanda    Solomon        1.48             317.04   
54638     Dwayne       Bell        3.01            1680.61   

       profit_of_each_line  
0                 35004.45  
1                   192.16  
2                   203.94  
3                   323.23  
4                  1274.30  
...                    ...  
54634              1017.99  
54635              1346.11  
54636              3924.65  
54637              -151.28  
54638              1422.70  

[54639 rows x 5 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [157]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]
df_filtered = df[df['order_id'].isin(order_ids)]
df_filtered
total_price_per_order = df_filtered.groupby('order_id')['total_price'].sum()
total_price_per_order



order_id
2173913    15713.63
2742071    17894.53
6128929    93053.30
Name: total_price, dtype: float64

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [158]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



In [159]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [160]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [161]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
